### PR TITLE
BlobDB: refactor DB open logic

### PR DESCRIPTION
**Pull request by yiwu-arbug**
Summary:
Refactor BlobDB open logic. List of changes:

Major:
* On reopen, mark blob files found as immutable, do not use them for writing new keys.
* Not to scan the whole file to find file footer. Instead just seek to the end of the file and try to read footer.

Minor:
* Move most of the real logic from blob_db.cc to blob_db_impl.cc.
* Not to hold shared_ptr of event listeners in global maps in blob_db.cc
* Some changes to BlobFile interface. 
* Improve logging and error handling.

Test Plan:
existing tests.

### CODE REVIEW

File: utilities/blob_db/blob_file.h [hunk 1]
```diff
@@ -98,19 +104,25 @@ class BlobFile {
 
   bool header_valid_;
 
+  bool footer_valid_;
```
Comment (sagar0) [inline utilities/blob_db/blob_file.h]: Wondering if these is a need for these to be made atomic?

File: utilities/blob_db/blob_file.h [hunk 2]
```diff
@@ -98,19 +104,25 @@ class BlobFile {
...
+           Logger* info_log);
 
   ~BlobFile();
 
   uint32_t column_family_id() const;
 
-  // Returns log file's pathname relative to the main db dir
-  // Eg. For a live-log-file = blob_dir/000003.blob
+  void SetColumnFamilyId(uint32_t cf_id) {
+    column_family_id_ = cf_id;
```
Comment (sagar0) [inline utilities/blob_db/blob_file.h]: We are not yet supporting column families in BlobDB, are we? 
Having this function somehow gave me the impression that we are already supporting column families.

Comment (yiwu-arbug) [reply utilities/blob_db/blob_file.h]: I'm leaving room for column family support.

File: utilities/blob_db/blob_file.h (hunk 1 above)
Comment (yiwu-arbug) [reply utilities/blob_db/blob_file.h]: I'll cleanup mutex usage and the BlobFile interface in the next PR.

**Comment by yiwu-arbug** [comment]
Thanks for review!

**Pull request closed without merging**